### PR TITLE
#5463: fix quick bar losing text selection when user types

### LINKED ISSUE #5463: Typing in the Quick Bar causes me to lose selectionText
Typing in the Quick Bar causes me to lose selectionText
From:
@Mike
Impact Details:
This is impacting all of the AI mods
Steps to Reproduce:
https://www.loom.com/share/4e7f32f07aa2400ab4f77e39682b8ef6

Request:
Typing in the Quick Bar causes me to lose selectionText

From:
<@U016Y2KJQ4T>

Impact Details:
This is impacting all of the AI mods

Steps to Reproduce:
<https://www.loom.com/share/4e7f32f07aa2400ab4f77e39682b8ef6|https://www.loom.com/share/4e7f32f07aa2400ab4f77e39682b8ef6>

[Slack Message](https://pixiebrix.slack.com/archives/C0436P48QHY/p1680390765442399)

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Fixes #5463 
- Wraps selection logic in useEffect so re-render doesn't cause the original selection to be lost

## Work Remaining

- [x] Figure out how to snapshot and send events to the portal
- [x] Fix broken tests - they work locally 🤷 

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer: @BLoe

### CODE REVIEW

File: src/components/quickBar/QuickBarApp.test.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  });
+
+  it("should render toggled", async () => {
+    render(<QuickBarApp />);
+
+    await act(async () => {
+      window.dispatchEvent(new Event(QUICKBAR_EVENT_NAME));
+
+      // Fast-forward until all timers have been executed
+      jest.advanceTimersByTime(2000);
```
Comment (BLoe) [inline src/components/quickBar/QuickBarApp.test.tsx:90]: Could also use `jest.runOnlyPendingTimers()` here probably.

File: src/components/quickBar/QuickBarApp.test.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    // Avoid timeout when using useFakeTimers: https://github.com/testing-library/user-event/issues/833
+    const user = userEvent.setup({ delay: null });
+
+    render(<QuickBarApp />);
+
+    await act(async () => {
+      window.dispatchEvent(new Event(QUICKBAR_EVENT_NAME));
+
+      // Fast-forward until all timers have been executed
+      jest.advanceTimersByTime(2000);
```
Comment (BLoe) [inline src/components/quickBar/QuickBarApp.test.tsx:109]: `jest.runOnlyPendingTimers()`

File: src/components/quickBar/QuickBarApp.test.tsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    expect(screen.getByTestId("quickBar")).toBeVisible();
+    expect(screen.getByRole("combobox")).toBeVisible();
+    expect(saveSelectionMock).toHaveBeenCalledOnce();
+
+    await act(async () => {
+      await user.type(screen.getByRole("combobox"), "test");
+
+      // Fast-forward until all timers have been executed
+      jest.advanceTimersByTime(2000);
```
Comment (BLoe) [inline src/components/quickBar/QuickBarApp.test.tsx:120]: `jest.runOnlyPendingTimers()`

**Review by BLoe — COMMENTED** [review]
I think you need to call `jest.useRealTimers()` at the end of your tests but I actually have no idea if it carries over into other test files or not.

File: src/components/quickBar/QuickBarApp.test.tsx (hunk 2 above)
Comment (twschiller) [reply src/components/quickBar/QuickBarApp.test.tsx:120]: Added a comment on why I don't think runOnlyPendingTimers works in this case

**Comment by twschiller** [comment]
@BLoe thanks - this should be ready for review

**Review by BLoe — APPROVED** [review]

**Pull request merged**